tests/src/Integration/Engine/PrepareMetadataTest.php

```diff
@@ -0,0 +1,94 @@
+use Drupal\rules\Engine\RulesComponent;
+use Drupal\Tests\rules\Integration\RulesEntityIntegrationTestBase;
+
+/**
+ * Tests that the setup of the execution metadata state for an expression works.
+ *
+ * @group rules
+ */
+class PrepareMetadataTest extends RulesEntityIntegrationTestBase {
+
+  /**
+   * Tests that a variable can be added by an action and is then available.
+   */
+  public function testAddingVariable() {
+    $rule = $this->rulesExpressionManager->createRule();
+    $rule->addAction('rules_variable_add', ContextConfig::create()
+      ->setValue('type', 'string')
+      ->setValue('value', '')
+      ->provideAs('variable_added', 'result')
+    );
+
+    $state = ExecutionMetadataState::create();
+    $found = $rule->prepareExecutionMetadataState($state);
+    $this->assertTrue($state->hasDataDefinition('result'));
+    $this->assertTrue($found);
+  }
+
+  /**
+   * Tests partial state setup until an expression is reached in the tree.
+   */
+  public function testPreparingUntil() {
+    // Setup a rule with 2 actions.
+    $rule = $this->rulesExpressionManager->createRule();
+    $rule->addAction('rules_variable_add', ContextConfig::create()
+      ->setValue('type', 'string')
+      ->setValue('value', '')
+      ->provideAs('variable_added', 'result1')
+    );
+    $second_action = $this->rulesExpressionManager->createAction('rules_variable_add')
+      ->setConfiguration(ContextConfig::create()
+        ->setValue('type', 'string')
+        ->setValue('value', '')
+        ->provideAs('variable_added', 'result2')
+        ->toArray()
+      );
+    $rule->addExpressionObject($second_action);
+
+    $state = ExecutionMetadataState::create();
+    // Preparing the state until the second action means the variable of the
+    // first action is available, but the second is not yet.
+    $found = $rule->prepareExecutionMetadataState($state, $second_action);
+    $this->assertTrue($state->hasDataDefinition('result1'));
+    $this->assertFalse($state->hasDataDefinition('result2'));
+    $this->assertTrue($found);
+  }
+
+  /**
+   * Tests that state preparation also works for actions in a loop.
+   */
+  public function testPrepareInLoop() {
+    $rule = $this->rulesExpressionManager->createRule();
+
+    $loop = $this->rulesExpressionManager->createInstance('rules_loop', ['list' => 'string_list']);
+    $action = $this->rulesExpressionManager->createAction('rules_test_string')
+      ->setConfiguration(ContextConfig::create()
+        ->setValue('text', 'x')
+        ->toArray()
+      );
+    $loop->addExpressionObject($action);
+
+    $rule->addExpressionObject($loop);
+
+    $state = RulesComponent::create($rule)
+      ->addContextDefinition('string_list', ContextDefinition::create('string')->setMultiple())
+      ->getMetadataState();
+
+    $found = $rule->prepareExecutionMetadataState($state, $action);
+    $this->assertTrue($state->hasDataDefinition('list_item'));
+    $this->assertTrue($found);
```

Comment:
Test coverage for that list item after the loop has been added.

We have a test case for loops (LoopTest), so we could add it there. Currently the tests are arranged by topic (integrity check, auto saving, RulesComponent usage), but also per expression (LoopTest and the unit tests for actions sets and condition sets).

I'm not sure which is better, since only one test class for an expression could also grow quite a lot - I like the topical tests.

Leaving this as is for now.